resolver/src/dns_lru.rs

```diff
@@ -37,40 +38,97 @@ impl LruValue {
+    /// A maximum TTL value for negative (`NXDOMAIN`) responses.
+    ///
+    /// `NXDOMAIN` responses with TTLs over `negative_max_ttl` will use
+    /// `negative_max_ttl` instead.
+    ///
+    ///  If this value is not set on the `TtlConfig` used to construct this
+    /// `DnsLru`, it will default to [`MAX_TTL`] seconds.
+    ///
+    /// [`MAX_TTL`]: const.MAX_TTL.html
+    negative_max_ttl: Duration,
+
 }
 
 #[derive(Copy, Clone, Debug, Default)]
 pub(crate) struct TtlConfig {
     /// An optional minimum TTL value for positive responses.
     ///
-    /// Positive responses with TTLs under `min_positive_ttl` will use
-    /// `` instead.
-    pub min_positive_ttl: Option<Duration>,
+    /// Positive responses with TTLs under `positive_min_ttl` will use
+    /// `positive_min_ttl` instead.
+    pub positive_min_ttl: Option<Duration>,
     /// An optional minimum TTL value for negative (`NXDOMAIN`) responses.
     ///
-    /// `NXDOMAIN` responses with TTLs under `min_negative_ttl` will use
-    /// `min_negative_ttl` instead.
-    pub min_negative_ttl: Option<Duration>,
+    /// `NXDOMAIN` responses with TTLs under `negative_min_ttl will use
+    /// `negative_min_ttl` instead.
+    pub negative_min_ttl: Option<Duration>,
+    /// An optional maximum TTL value for positive responses.
+    ///
+    /// Positive responses with TTLs positive `positive_max_ttl` will use
+    /// `positive_max_ttl` instead.
+    pub positive_max_ttl: Option<Duration>,
+    /// An optional maximum TTL value for negative (`NXDOMAIN`) responses.
+    ///
+    /// `NXDOMAIN` responses with TTLs over `negative_max_ttl` will use
+    /// `negative_max_ttl` instead.
+    pub negative_max_ttl: Option<Duration>,
+}
+
+impl TtlConfig {
+    pub(crate) fn from_opts(opts: &config::ResolverOpts) -> TtlConfig {
+        TtlConfig {
+            positive_min_ttl: opts.positive_min_ttl,
+            negative_min_ttl: opts.negative_min_ttl,
+            positive_max_ttl: opts.positive_max_ttl,
+            negative_max_ttl: opts.negative_max_ttl,
+        }
+    }
 }
 
 impl DnsLru {
     pub(crate) fn new(capacity: usize, ttl_cfg: TtlConfig) -> Self {
-        let TtlConfig { min_positive_ttl, min_negative_ttl } = ttl_cfg;
+        let TtlConfig {
+            positive_min_ttl,
+            negative_min_ttl,
+            positive_max_ttl,
+            negative_max_ttl,
+        } = ttl_cfg;
         let cache = LruCache::new(capacity);
         Self {
             cache,
-            min_positive_ttl: min_positive_ttl.unwrap_or_else(|| Duration::from_secs(0)),
-            min_negative_ttl: min_negative_ttl.unwrap_or_else(|| Duration::from_secs(0)),
+            positive_min_ttl: positive_min_ttl.unwrap_or_else(|| Duration::from_secs(0)),
+            negative_min_ttl: negative_min_ttl.unwrap_or_else(|| Duration::from_secs(0)),
+            positive_max_ttl: positive_max_ttl
+                .unwrap_or_else(|| Duration::from_secs(MAX_TTL as u64)),
```

Comment:
Putting this comment here, but it probably belongs on the type definition on the fields in the Type. This case made me realize a potential issue we have to concern ourselves with: https://github.com/bluejekyll/trust-dns/blob/master/proto/src/rr/resource.rs#L69

The TTL field on Records is a `u32::MAX_VALUE`. We just need to think through the implications of this. I like having Duration, but it's potential range is greater than the value acceptable in DNS (I missed this in previous reviews...). I think the current usage is probably fine, given that these are never sent on the wire... but it does mean that a larger than standard TTL can be used for these defaults.

Reply:
This is a good point, thanks for pointing that out! 

I suppose if a max value is set that's greater than `u32::MAX_VALUE`, we're _technically_ behaving correctly, since any received TTLs would be, by definition, under the limit. I

 suppose this could be documented, or we could change the API to accept these configurations as `u32`s, which _does_ seem like the best way of making it clear to the user.

Reply:
In Conduit, we are parsing `Duration`s from environment variables, so part of my motivation was for this API to take `Duration`s as it makes my life easier when I actually go to use it downstream. However, if you think using `u32`s is more correct, we can always convert them back into a number of seconds in `u32` form.

Thinking about it, I feel like it's probably a more common user assumption to expect this kind of API in general to be configured using `Duration`s instead of integers. However, given that the limitation is due to the underlying DNS protocol, `u32`s might still be more correct... I can see it either way. What do you think?

Reply:
Yes, I agree with duration. I considered removing my comment before posting it, but decided it was worth it if for no other reason than the thought process.

It’s probably worth documenting this, just to be pedantic, and hopefully stop anyone from potentially using it on the wire, though that would be an odd case.

Reply:
> It’s probably worth documenting this, just to be pedantic, and hopefully stop anyone from potentially using it on the wire, though that would be an odd case.

Yeah, that sounds like the right thing to do here. Will add it to the documentation!